### PR TITLE
En 6778 more waiting than eligible

### PR DESCRIPTION
**Pull request by raduchis**
If the waiting list contained more nodes than the eligible list, at the beginning of the next epoch, all waiting nodes would be moved to the eligible lists, leading to a possible increase in the number of eligible nodes

- fix if waiting contains more than eligible, by just adding nodes from the waiting list to the eligible to fill the remaining space
- fixed Less() function in validator sorting to sort indexes ascending

### CODE REVIEW

File: sharding/hashValidatorShuffler.go [hunk 1]
```diff
@@ -533,6 +533,39 @@ func moveNodesToMap(destination map[uint32][]Validator, source map[uint32][]Vali
 	return nil
 }
 
+// moveMaxNumNodesToMap moves the validators in the source list to the corresponding destination list
+// but adding just enough nodes so that at most the number of nodes is kept in the destination list
+func moveMaxNumNodesToMap(destination map[uint32][]Validator, source map[uint32][]Validator, numMeta, numShard uint32) error {
```
Comment (sasurobert) [inline sharding/hashValidatorShuffler.go:538]: arrange variables. put each in new line

File: sharding/hashValidatorShuffler.go [hunk 2]
```diff
@@ -533,6 +533,39 @@ func moveNodesToMap(destination map[uint32][]Validator, source map[uint32][]Vali
...
+		maxNodes := numShard
+		if k == core.MetachainShardId {
+			maxNodes = numMeta
+		}
+
+		numNeededNodes := computeNeededNodes(destination[k], source[k], maxNodes)
+		destination[k] = append(destination[k], v[0:numNeededNodes]...)
+		source[k] = v[numNeededNodes:]
+	}
+	return nil
```
Comment (sasurobert) [inline sharding/hashValidatorShuffler.go:553]: add empty line before

File: sharding/hashValidatorShuffler.go [hunk 3]
```diff
@@ -533,6 +533,39 @@ func moveNodesToMap(destination map[uint32][]Validator, source map[uint32][]Vali
...
+	numNeededNodes := uint32(0)
+	numCurrentNodes := uint32(len(destination))
+	numSourceNodes := uint32(len(source))
+	if maxNumNodes > numCurrentNodes {
+		numNeededNodes = maxNumNodes - numCurrentNodes
+	}
+	if numSourceNodes < numNeededNodes {
+		return numSourceNodes
+	}
+	return numNeededNodes
```
Comment (sasurobert) [inline sharding/hashValidatorShuffler.go:566]: add empty line before

File: sharding/indexHashedNodesCoordinator.go [hunk 4]
```diff
@@ -46,9 +46,9 @@ func (v validatorList) Swap(i, j int) { v[i], v[j] = v[j], v[i] }
 // Sorting of validators should be by index and public key
 func (v validatorList) Less(i, j int) bool {
 	if v[i].Index() == v[j].Index() {
-		return bytes.Compare(v[i].PubKey(), v[j].PubKey()) > 0
+		return bytes.Compare(v[i].PubKey(), v[j].PubKey()) < 0
```
Comment (sasurobert) [inline sharding/indexHashedNodesCoordinator.go:49]: this was maybe the problem for stake unstake.

File: sharding/hashValidatorShuffler.go [hunk 5]
```diff
@@ -533,6 +533,39 @@ func moveNodesToMap(destination map[uint32][]Validator, source map[uint32][]Vali
...
 }
 
+// moveMaxNumNodesToMap moves the validators in the source list to the corresponding destination list
+// but adding just enough nodes so that at most the number of nodes is kept in the destination list
+func moveMaxNumNodesToMap(destination map[uint32][]Validator, source map[uint32][]Validator, numMeta, numShard uint32) error {
+	if destination == nil {
+		return ErrNilOrEmptyDestinationForDistribute
+	}
+
+	for k, v := range source {
```
Comment (SebastianMarian) [inline sharding/hashValidatorShuffler.go:543]: shardID, validator ?

File: sharding/hashValidatorShuffler.go (hunk 1 above)
Comment (raduchis) [reply sharding/hashValidatorShuffler.go:538]: fixed

File: sharding/hashValidatorShuffler.go (hunk 2 above)
Comment (raduchis) [reply sharding/hashValidatorShuffler.go:553]: fixed

File: sharding/hashValidatorShuffler.go (hunk 3 above)
Comment (raduchis) [reply sharding/hashValidatorShuffler.go:566]: fixed

File: sharding/hashValidatorShuffler.go (hunk 5 above)
Comment (raduchis) [reply sharding/hashValidatorShuffler.go:543]: fixed

**Pull request merged**